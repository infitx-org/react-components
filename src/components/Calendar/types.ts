export type PossibleDate = Date | undefined;

export type PossibleDay = string | undefined;

export type Matrix = PossibleDay[][];

export type DisabledDays = (day: Date) => boolean;

export type DateRange = [PossibleDate, PossibleDate];
