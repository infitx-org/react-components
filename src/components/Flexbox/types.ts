export enum TopBottom {
  "top" = "flex-start",
  "bottom" = "flex-end",
}
export enum LeftRight {
  "left" = "flex-start",
  "right" = "flex-end",
}

export enum Mappers {
  "top" = "flex-start",
  "bottom" = "flex-end",
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap<T extends string> = `${AlignItems}` | `${T}`;
export type JustifyWithMap<T extends string> = `${JustifyContent}` | `${T}`;
