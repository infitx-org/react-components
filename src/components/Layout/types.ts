export enum AlignMap {
  "top" = "flex-start",
  "bottom" = "flex-end",
}

export enum JustifyMap {
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap = `${AlignItems | keyof typeof AlignMap}`;
export type JustifyWithMap = `${JustifyContent | keyof typeof JustifyMap}`;
