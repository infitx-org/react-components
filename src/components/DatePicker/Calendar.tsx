import React from "react";
import DayPicker from "react-day-picker";
import classnames from "classnames";
import useOverlayPosition from "hooks/useOverlayPosition";
import "react-day-picker/lib/style.css";
import "./DayPicker.scss";

interface CalendarProps {
  selectedDate: Date | undefined;
  onDayClick: (day: Date) => void;
}

export default function Calendar({ selectedDate, onDayClick }: CalendarProps) {
  const ref = React.useRef<HTMLDivElement>(null);
  const { top, bottom, height, reverse } = useOverlayPosition<HTMLDivElement>(
    ref.current
  );
  const className = classnames([
    "rc-datepicker__calendar",
    reverse && "rc-datepicker__calendar--reverse",
  ]);
  return (
    <div
      ref={ref}
      onClick={(e) => e.stopPropagation()}
      className={className}
      role="presentation"
      style={{ top, bottom, maxHeight: height }}
    >
      <div className="rc-datepicker__calendar__container">
        <DayPicker selectedDays={selectedDate} onDayClick={onDayClick} />
      </div>
    </div>
  );
}
