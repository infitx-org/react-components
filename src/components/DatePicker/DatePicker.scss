@use "../../scss/base/colors.scss";
@use "../../scss/mixins/field.scss";
@use "../../scss/mixins/any.scss";

.rc-datepicker {
  @include any.rc-component;
  @include field.input;

  flex: 2 1 auto;
  vertical-align: middle;
  position: relative;
  height: 20px;
  background: transparent;
  border: none;
  outline: none;
  padding: 0;
  pointer-events: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: colors.$dark-gray;
}
