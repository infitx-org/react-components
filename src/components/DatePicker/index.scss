@import "./DatePicker.scss";
@import "react-day-picker/lib/style.css";
@import "./components/Calendar.scss";
@import "./components/DayPicker.scss";