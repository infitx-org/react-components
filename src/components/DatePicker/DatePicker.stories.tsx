import { Story } from "@storybook/react";
import log from "resources/log";
import DatePicker, { DatePickerProps } from "./DatePicker";

export default {
  title: "Components/DatePicker",
  component: DatePicker,
};

const Template: Story<DatePickerProps> = (args) => <DatePicker {...args} />;

export const Default = Template.bind({});
Default.args = {
  kind: "primary",
  size: "large",
  value: undefined,
  className: undefined,
  placeholder: "Choose a date",
  disabled: false,
  required: false,
  invalid: false,
  pending: false,
  withTime: false,
  onChange: log,
};

export const SelectedDate = Template.bind({});
SelectedDate.args = {
  value: new Date().toISOString(),
};

export const WithTime = Template.bind({});
WithTime.args = {
  value: new Date().toISOString(),
  withTime: true,
};

export const Format = Template.bind({});
Format.args = {
  format: "MM/dd/yyyy",
  value: new Date().toISOString(),
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Placeholder = Template.bind({});
Placeholder.args = {
  placeholder: "Pick a date",
};

export const Pending = Template.bind({});
Pending.args = {
  pending: true,
};

export const Required = Template.bind({});
Required.args = {
  required: true,
};

export const Invalid = Template.bind({});
Invalid.args = {
  invalid: true,
};

export const Small = Template.bind({});
Small.args = {
  size: "small",
};

export const Medium = Template.bind({});
Medium.args = {
  size: "medium",
};

export const Large = Template.bind({});
Large.args = {
  size: "large",
};
