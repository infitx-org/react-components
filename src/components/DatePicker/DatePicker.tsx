import React from "react";
import classnames from "classnames";
import { InputSize, KeyCode } from "types";
import { format as dateFormat } from "date-fns";
import mergeRefs from "utils/mergeRefs";
import Field, { Loader, Placeholder, InvalidIcon } from "components/Field";
import Calendar from "./components/Calendar";
import CalendarIcon from "./components/CalendarIcon";
import "./DatePicker.scss";

type DateValue = Date | undefined;

export interface DatePickerProps
  extends Omit<
    React.InputHTMLAttributes<HTMLInputElement>,
    "size" | "onChange"
  > {
  className?: string;
  size?: `${InputSize}`;
  format?: string;
  placeholder?: string;
  value?: string;
  required?: boolean;
  invalid?: boolean;
  pending?: boolean;
  onChange?: (date: DateValue) => void;
}

export default React.forwardRef(function DatePicker(
  {
    className,
    size = InputSize.Large,
    format = "MMM do yyyy, HH:mm:ss",
    placeholder,
    value,
    required,
    invalid,
    pending,
    onChange,
    ...props
  }: DatePickerProps,
  ref: React.ForwardedRef<HTMLInputElement>
): JSX.Element {
  function getDateFromString(date?: string): DateValue {
    return date ? new Date(date) : undefined;
  }

  const inputRef = React.useRef<HTMLInputElement>(null);
  const [selectedDate, setDate] = React.useState<DateValue>(
    getDateFromString(value)
  );
  const [open, setOpen] = React.useState(false);
  const [focused, setFocused] = React.useState(false);

  React.useEffect(() => {
    setDate(getDateFromString(value));
  }, [value]);

  function getStringFromDate(date?: Date): string {
    if (date) {
      return dateFormat(date, format);
    }
    return "";
  }

  function enter() {
    setFocused(true);
    setOpen(true);
    inputRef.current?.focus();
  }

  function leave() {
    setFocused(false);
    setOpen(false);
  }

  function onDayClick(day: Date, selected?: boolean) {
    const newDate = selected ? undefined : day;
    setDate(newDate);
    inputRef.current?.focus();
    onChange?.(newDate);
  }

  function onFocus(e: React.FocusEvent<HTMLInputElement>) {
    if (!focused) {
      props.onFocus?.(e);
      enter();
    }
  }

  function onBlur(e: React.FocusEvent<HTMLInputElement>) {
    if (!focused) {
      props.onBlur?.(e);
    }
  }

  function onFieldClick(e: React.MouseEvent<HTMLDivElement>): void {
    e.preventDefault();
    e.stopPropagation();
    if (e.target === inputRef.current) {
      return;
    }
    if (open) {
      leave();
    } else {
      enter();
    }
  }

  function onKeyDown(e: React.KeyboardEvent<HTMLInputElement>) {
    const { keyCode } = e;

    if (keyCode === KeyCode.Tab) {
      leave();
      return;
    }
    if (keyCode === KeyCode.Return) {
      e.preventDefault();
      if (open) {
        setOpen(false);
      } else {
        setOpen(true);
      }
    }
  }

  const inputClassName = classnames([
    "rc-datepicker",
    `rc-datepicker--${size}`,
  ]);

  const visibleValue = getStringFromDate(selectedDate);
  return (
    <Field
      className={className}
      size={size}
      required={required && selectedDate === undefined}
      pending={pending}
      invalid={invalid}
      disabled={props.disabled}
      focused={focused}
      onClick={onFieldClick}
      onClickOutside={leave}
    >
      {placeholder && (
        <Placeholder
          label={placeholder}
          active={!!selectedDate || focused}
          size={size}
        />
      )}
      <input
        {...props}
        className={inputClassName}
        type="text"
        ref={mergeRefs<HTMLInputElement>(ref, inputRef)}
        onChange={(e) => e.preventDefault()}
        onFocus={onFocus}
        onBlur={onBlur}
        onKeyDown={onKeyDown}
        value={visibleValue}
        readOnly
      />
      {pending && <Loader size={size} />}
      {invalid && <InvalidIcon size={size} />}
      <CalendarIcon size={size} />
      {open && <Calendar selectedDate={selectedDate} onDayClick={onDayClick} />}
    </Field>
  );
});
