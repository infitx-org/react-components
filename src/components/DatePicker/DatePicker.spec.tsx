import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { InputSize } from "types";
import DatePicker from "./DatePicker";

const commonProps = {
  onChange: jest.fn(),
};

function getCalendarDays(container: HTMLElement): NodeListOf<HTMLDivElement> {
  return container.querySelectorAll(
    ".DayPicker-Day:not(.DayPicker-Day--outside)"
  ) as NodeListOf<HTMLDivElement>;
}
function getCalendar(container: HTMLElement): HTMLDivElement {
  return container.querySelector(".rc-datepicker__calendar") as HTMLDivElement;
}
function getInput(container: HTMLElement): HTMLInputElement {
  return container.querySelector("input[type='text']") as HTMLInputElement;
}

describe("tests the datepicker props", () => {
  it("renders the datepicker", () => {
    const { container } = render(<DatePicker {...commonProps} />);
    expect(container.querySelector(".rc-datepicker")).toBeTruthy();
    expect(container.querySelectorAll('input[type="text"]')).toHaveLength(1);
  });

  it("does not render the closed calendar", () => {
    const { container } = render(<DatePicker {...commonProps} />);
    expect(container.querySelector(".rc-datepicker__calendar")).toBeFalsy();
  });

  it("renders the date", () => {
    const { container } = render(
      <DatePicker
        {...commonProps}
        value={new Date("12/31/2021").toString()}
        format="MM/dd/yyyy"
      />
    );
    const input = getInput(container);
    expect(input.value).toBe("12/31/2021");
  });

  it("renders no date when value is not specified", () => {
    const { container } = render(<DatePicker {...commonProps} />);
    const input = getInput(container);
    expect(input.value).toBe("");
  });

  it("renders the placeholder", () => {
    const { container } = render(
      <DatePicker {...commonProps} placeholder="test" />
    );
    expect(container.querySelector(".placeholder")).toBeTruthy();
  });

  it("renders the classname", () => {
    const { container } = render(
      <DatePicker {...commonProps} className="test" />
    );
    expect(container.querySelector(".test")).toBeTruthy();
  });

  it("renders as disabled", () => {
    const { container } = render(<DatePicker {...commonProps} disabled />);
    const input = getInput(container);
    expect(input.disabled).toBeTruthy();
  });

  it("renders as pending", () => {
    const { container } = render(<DatePicker {...commonProps} pending />);
    expect(container.querySelector(".rc-spinner")).toBeTruthy();
  });

  it("renders as required", () => {
    const { container } = render(<DatePicker {...commonProps} required />);
    expect(container.querySelector(".rc-field--required")).toBeTruthy();
  });

  it("renders as invalid", () => {
    const { container } = render(<DatePicker {...commonProps} invalid />);
    expect(container.querySelector(".rc-field--invalid")).toBeTruthy();
    expect(container.querySelector(".rc-invalid-icon")).toBeTruthy();
  });

  it("renders the small, medium, large sizes", () => {
    Object.values(InputSize).forEach((size) => {
      const { container } = render(<DatePicker {...commonProps} size={size} />);
      expect(container.querySelector(`.rc-field--${size}`)).toBeTruthy();
    });
  });

  it("renders the calendar when clicked", () => {
    const { container } = render(<DatePicker {...commonProps} />);
    userEvent.click(getInput(container));
    expect(getCalendar(container)).toBeInTheDocument();
  });

  it("renders the calendar when focused", () => {
    const { container } = render(<DatePicker {...commonProps} />);
    fireEvent.focus(getInput(container));
    expect(getCalendar(container)).toBeInTheDocument();
  });

  it("selects a date when clicking a day", () => {
    const { container } = render(<DatePicker {...commonProps} format="dd" />);
    userEvent.click(getInput(container));
    const days = getCalendarDays(container);
    userEvent.click(days[0]);
    expect(getInput(container).value).toBe("01");
  });

  it("unselects a date when clicking the same day", () => {
    const { container } = render(<DatePicker {...commonProps} format="dd" />);
    userEvent.click(getInput(container));
    const days = getCalendarDays(container);
    userEvent.click(days[0]);
    userEvent.click(days[0]);
    expect(getInput(container).value).toBe("");
  });

  it("triggers onFocus when focusing", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <DatePicker {...commonProps} onFocus={mockFn} />
    );
    fireEvent.focus(getInput(container));
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onFocus when clicking", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <DatePicker {...commonProps} onFocus={mockFn} />
    );
    userEvent.click(getInput(container));
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onFocus when tabbing", () => {
    const mockFn = jest.fn();
    render(<DatePicker {...commonProps} onFocus={mockFn} />);
    userEvent.tab();
    expect(mockFn).toHaveBeenCalled();
  });

  it("triggers onBlur when clicking out", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <DatePicker {...commonProps} onBlur={mockFn} />
    );
    userEvent.click(getInput(container));
    userEvent.click(document.body);
    expect(mockFn).toHaveBeenCalled();
  });

  // it("triggers onBlur when tabbing away", () => {
  //   const mockFn = jest.fn();
  //   const { container } = render(<DatePicker {...commonProps} onBlur={mockFn} />);
  //   userEvent.click(getInput(container));
  //   userEvent.tab();
  //   expect(mockFn).toHaveBeenCalled();
  // });

  it("triggers onChange when selecting a date", () => {
    const mockEvent = jest.fn();
    const { container } = render(<DatePicker onChange={mockEvent} />);
    userEvent.click(getInput(container));
    const days = getCalendarDays(container);
    userEvent.click(days[0]);
    // const [call] = mockEvent.calls;
    expect(mockEvent).toHaveBeenCalled();
    const [[date]] = mockEvent.mock.calls;
    expect(date).toBeInstanceOf(Date);
  });

  it("triggers onChange when clicking the same day", () => {
    const mockFn = jest.fn();
    const { container } = render(
      <DatePicker
        {...commonProps}
        onChange={mockFn}
        value={new Date("01/01/2021").toString()}
      />
    );
    userEvent.click(getInput(container));
    const days = getCalendarDays(container);
    userEvent.click(days[0]);
    expect(mockFn).toHaveBeenCalledWith(undefined);
  });
});

// Snapshot testing
it("renders the DatePicker correctly when multiple props are set", () => {
  const { container } = render(
    <DatePicker {...commonProps} placeholder="test" />
  );
  expect(container).toMatchSnapshot();
});
