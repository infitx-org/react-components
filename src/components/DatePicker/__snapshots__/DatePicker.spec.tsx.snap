// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the DatePicker correctly when multiple props are set 1`] = `
<div>
  <div
    class="rc-field rc-field--primary rc-field--large"
    role="presentation"
  >
    <label
      class="placeholder placeholder--large"
    >
      test
    </label>
    <input
      class="rc-datepicker rc-datepicker--large"
      readonly=""
      type="text"
      value=""
    />
    <svg
      class="rc-datepicker__calendar-icon"
      fill="#000"
      height="16"
      stroke="transparent"
      width="16"
    />
  </div>
</div>
`;
